### PR TITLE
NEXTEUROPA-12634: Add logic for blocking the change tracking in nexte…

### PR DESCRIPTION
**Pull request by gillesdeudon**
need drush "updatedb" and "cc all"
The rest is in the nexteuropa_editorial readme file.

### CODE REVIEW

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+    $tracking_lang[$current_lang] = $current_lang;
+  }
+  // We remove LANGUAGE_NONE because it is common to all translations.
+  unset($tracking_lang[LANGUAGE_NONE]);
+
+  $count = count($tracking_lang);
+  switch ($count) {
+    case 0:
+      // LANGUAGE_NONE case.
+      return t('The form cannot be saved because of tracked changes existing in this revision.');
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Maybe we could try something bit more simpler :) Below you will find my proposal. 
'Current content includes unprocessed changes in this revision. Please accept or reject them in order to save progress.'

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+    case 1:
+      $implied_lang = end($tracking_lang);
+      $lang_name = locale_language_name($implied_lang);
+      $message_param = array('@lang' => $lang_name);
+      return t('The form cannot be saved because of tracked changes existing in the @lang version.',
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Similar. Could be:
'Current content includes unprocessed changes in the @lang version. Please accept or reject them in order to save progress.'

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+        $message_param);
+
+    default:
+      $language_names = array();
+      foreach ($tracking_lang as $lang) {
+        $language_names[$lang] = locale_language_name($lang);
+      }
+      $message_param = array('@lang' => implode(', ', $language_names));
+      return t('The form cannot be saved because of tracked changes existing in the @lang versions.',
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Same stuff here.

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+Configuration > Content authoring > MULTISITE WYSIWYG > WYSIWYG workflow settings
+(path: "admin/config/content/multisite_wysiwyg/workbench_en");
+
+# Remark for editors: Tracking changes
+
+Depending on settings made in "WYSIWYG workflow settings", content saving can be
+blocked if tracked changes managed by the module "CKEditor LITE" module exist in
+"WYSIWYG" fields of this content or in one of these translations.
+
+In order to unblock the saving,editors must validate or refuse tracked changes
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: Missing space between `aving,editors`maybe something like this could be an improvement:
"In order to save content for configured workflow statuses, editors must accept or reject tracked changes if those are present."

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+It is also possible to set if content types that have not workbench moderation
+activated or if no states are selected below ("Block if status" is true checkbox).
+
+They can be modified on this administration page:
+Configuration > Content authoring > MULTISITE WYSIWYG > WYSIWYG workflow settings
+(path: "admin/config/content/multisite_wysiwyg/workbench_en");
+
+# Remark for editors: Tracking changes
+
+Depending on settings made in "WYSIWYG workflow settings", content saving can be
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: Maybe something like this:
Depending on the feature configuration options for given content workflow status (ex. published), saving progress could be blocked if tracked changes are not validated in current revision or one of translations.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+ * Field widget form alter callback.
+ *
+ * It inject a specific validate callback in achrge of scanning field value.
+ *
+ * @see nexteuropa_editorial_field_widget_text_textarea_form_alter()
+ * @see nexteuropa_editorial_field_widget_text_textarea_with_summary_form_alter()
+ * @see nexteuropa_editorial_field_widget_text_textfield_form_alter()
+ */
+function _nexteuropa_editorial_field_widget_tracking_form_alter(&$element, &$form_state, $context) {
+  // So far, only nodes are supported.
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: If so in following lines there should be checkpoint for 'node' type otherwise this function will be triggered all the time.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 7]
```diff
@@ -0,0 +1,429 @@
+<?php
+/**
+ * @file
+ * Code specific to customization for the change tracking management.
+ */
+
+/**
+ * Field widget form alter callback.
+ *
+ * It inject a specific validate callback in achrge of scanning field value.
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Maybe like this: (there is also typo `in charge of`)
Injects a custom validation callback which is searching for change tracking markup in the field value.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+ * @param object $entity
+ *   The entity to treat.
+ * @param string $type
+ *   The entity type.
+ *
+ * @throws EntityMalformedException
+ *   If the entity bundle is not set.
+ */
+function _nexteuropa_editorial_set_tracking_flags($entity, $type = 'node') {
+  // Tracking changes are only present in these "fields type".
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Maybe:
in the following "field types" (like in beneath variable name).

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+  $screened_field_types = array(
+    'text_textarea',
+    'text_textarea_with_summary',
+    'text_textfield',
+  );
+  list(,, $bundle) = entity_extract_ids($type, $entity);
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Do we need to call that function. For nodes bundle property should be already in $entity object.

Comment (gillesdeudon) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: It is to prepare the next step when we will open to other entity.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+ *
+ * @param object $entity
+ *   The entity to check.
+ * @param bool|FALSE $current_wbm_state
+ *   The entity current workbench moderation state (node only).
+ *
+ * @return bool
+ *   TRUE if the saving must be blocked; otherwise FALSE.
+ */
+function _nexteuropa_editorial_tracking_saving_to_block($entity, $current_wbm_state = FALSE) {
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Name of this function suggest that something is going to be saved to the block. Maybe:
_nexteuropa_editorial_tracking_changes_check()

File: profiles/common/modules/features/nexteuropa_editorial/nexteuropa_editorial.module [hunk 11]
```diff
@@ -114,3 +132,94 @@ function nexteuropa_editorial_module_implements_alter(&$implementations, $hook)
...
+ */
+function nexteuropa_editorial_field_widget_text_textfield_form_alter(&$element, &$form_state, $context) {
+  _nexteuropa_editorial_field_widget_tracking_form_alter($element, $form_state, $context);
+}
+
+/**
+ * Implements hook_entity_load().
+ */
+function nexteuropa_editorial_entity_load($entities, $type) {
+  foreach ($entities as $entity) {
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/nexteuropa_editorial.module]: This faction is really used often. I would suggest to put here check for the `node` type and also I would recommend to use the arg() function to attach it only to edit node pages.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 12]
```diff
@@ -0,0 +1,429 @@
...
+  if ($entity && isset($entity->entity_type)) {
+    $type = $entity->entity_type;
+    $cached_settings = &drupal_static(__FUNCTION__ . '_' . $type);
+
+    // So far, only nodes are supported.
+    if (empty($cached_settings) && ($type == 'node')) {
+      $status = variable_get('nexteuropa_editorial_tracking_status');
+      $states = variable_get('nexteuropa_editorial_tracking_wbm_states', array());
+      $states_to_check = array();
+      foreach ($states as $state_name => $state_status) {
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Do we need to perform this loop. It is just removing key and creates new array just with values.
Maybe http://php.net/manual/en/function.array-values.php could be an alternative for that.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+  if (isset($entity->translations)) {
+    // Entity translation case.
+    $available_transl = $entity->translations;
+    $available_transl_langs = array_keys($available_transl->data);
+  }
+
+  $ent_field_instances = field_info_instances($type, $bundle);
+  // Note: special field types like field collection or paragraphs
+  // are not supported yet.
+  foreach ($ent_field_instances as $field_name => $ent_field_instance) {
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Here I would suggest to check widget type $ent_field_instances['field_ne_body']['widget']['type']. It will save some code later on.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+
+      // Treating current revision.
+      $tracking_detected = FALSE;
+      if (isset($form_state['tracked_changed_detected'])) {
+        $tracking_detected = $form_state['tracked_changed_detected'];
+      }
+
+      if ($tracking_detected) {
+        foreach ($form_state['tracked_changed_detected'] as $field_path) {
+          form_set_error($field_path, t('The field contains tracked changes that may not be saved.'));
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: It would be good to pass field name to the message. Important for cases where you will have more WYSIWYG fields instances.

**Comment by netlooker** [comment]
👎

File: profiles/common/modules/features/nexteuropa_editorial/nexteuropa_editorial.module (hunk 11 above)
Comment (gillesdeudon) [inline profiles/common/modules/features/nexteuropa_editorial/nexteuropa_editorial.module]: Ok for checking the node type but not for the use of the arg() because I ensured that the type is present in the entity attached to the form. I tried to limit its use when I deal with forms because they can be embedded by custom or contrib modules into other pages.Then, I have no trust in it.

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc (hunk 13 above)
Comment (gillesdeudon) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: It is the case but in helper functions. I thought the widget type could be required by future implementations managed by the module. I did not see the same need for the"text_processing".

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 15]
```diff
@@ -222,22 +212,45 @@ function _nexteuropa_editorial_tracking_wbm_validate($form, &$form_state) {
...
+      // In this case, the inserted value is stored differently.
+      $value = $element['value']['#value'];
+      $element = $element['value'];
+    }
+
+    $field_path = implode('][', $element['#parents']);
+    unset($form_state['tracked_changed_detected'][$field_path]);
+    if (preg_match('/<span[^>]+class="[^"]*ice-[^>]+>/', $value)) {
+      // Flag used in entity validate callback.
+      $form_state['tracked_changed_detected'][$field_path] = TRUE;
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: Typo: `changed` - ['tracking_changes_detected']

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc [hunk 16]
```diff
@@ -222,22 +212,45 @@ function _nexteuropa_editorial_tracking_wbm_validate($form, &$form_state) {
...
+      if (!empty($element['#title'])) {
+        $title = $element['#title'];
+      }
+      else {
+        // Multi value fields have an empty '#title'.
+        $info = field_info_instance($entity->entity_type, $element['#field_name'], $element['#bundle']);
+        $title = check_plain($info['label']);
+      }
+
+      $message = t('!title contains tracked changes that may not be saved.',
```
Comment (netlooker) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: As we have also check on the entity level which will give a nice description what you need to do, I suggest to make this message bit more generic and descriptive ex. 'contains tracking changes information.'

**Comment by netlooker** [comment]
👍

File: profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc (hunk 13 above)
Comment (Fefaine) [inline profiles/common/modules/features/nexteuropa_editorial/includes/nexteuropa_editorial.tracking.inc]: I don't think we should limit this to one field, so i would not follow this remark. Or maybe I didn't understand?

File: tests/features/change_tracking_workflow.feature [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    Then I should see the error message "Tracked changes detected in the French version. To save progress, please accept or reject them, or change the content state."
+    When I select "Validated" from "state"
+    And I press "Apply"
+    Then I should see the error message "Tracked changes detected in the French version. To save progress, please accept or reject them, or change the content state."
+    When I select "Needs Review" from "state"
+    And I press "Apply"
+    Then I should not see the error message "Tracked changes detected in the French version. To save progress, please accept or reject them, or change the content state."
+
+    Examples:
+      | blocked                                                                                                                                                                                                                                  |
```
Comment (Fefaine) [inline tests/features/change_tracking_workflow.feature]: Just a remark for long strings, documentation offers the option to use PyStrings  
https://media.readthedocs.org/pdf/behat/v3.0/behat.pdf (p25)

**Comment by Fefaine** [comment]
:+1:

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 18]
```diff
@@ -0,0 +1,46 @@
+The "Editorial" feature supplies default settings and permissions linked to
+editorial features of the Next Europa platform based on "OG" and "Workbench moderation" modules.
+
+It also contains alterations of:
+* The automatically assignment of the "editorial team member" role to
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: The automatic assignment

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+# Installation
+
+The module is activated when the platform is installed.
+
+# Configuration
+
+The module by default sets some permissions and variables used by the different
+editorial features.
+So, it is possible to set workbench moderation states that are not possible if
+tracked changes exists in WYSIWYG fields ("Block the saving for these
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: tracked changes exist ...

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+Workbench Moderation states" checkboxes).
+It is also possible to set if content types that have not workbench moderation
+activated or if no states are selected below ("Block if status" is true checkbox).
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: "if content types that have not workbench moderation activated" is not clear to me, can you rephrase?

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+They can be modified on this administration page:
+Configuration > Content authoring > MULTISITE WYSIWYG > WYSIWYG workflow settings
+(path: "admin/config/content/multisite_wysiwyg/workbench_en");
+
+# Remark for editors: Tracking changes
+
+Depending on the feature configuration options for given content workflow status
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: for **a** given content workflow status

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+(ex. published), saving progress could be blocked if tracked changes are not
+validated in the current revision or one of translations.
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: ... or one of **the** translations

File: profiles/common/modules/features/nexteuropa_editorial/README.md [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+
+In order to save content for configured workflow statuses, editors must
+accept or reject tracked changes if those are present,or change the
```
Comment (cyberwolf) [inline profiles/common/modules/features/nexteuropa_editorial/README.md]: add a space before and after the comma

File: tests/features/change_tracking_workflow.feature [hunk 24]
```diff
@@ -0,0 +1,170 @@
+Feature: Change tracking features
+  In order to ease communications between editors while they manage website
+  content, a tracking feature is available with WYSIWYG fields.
+  I want to use it while I edit content, and I want to see tracked
+  changes while I view content draft.
```
Comment (cyberwolf) [inline tests/features/change_tracking_workflow.feature]: I think "draft content" is better instead of "content draft".

File: tests/features/change_tracking_workflow.feature [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+      | title            | Lorem ipsum dolor sit amet                                      |
+      | body             | <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p> |
+      | moderation state | draft                                                           |
+    And I click "Edit draft"
+    And I fill in "Body" with "<html>"
+    And I press "Save"
+    Then the response should contain "<expected>"
+    And I should see the following warning messages:
+      | warning messages | <strong>The change tracking is activated on some fields of this "Basic page" content</strong>.<br /> <small>Please accept or reject tracked changes before setting the content state to validated or published.</small> |
+    And I should see an "div.ICE-Tracking" element
```
Comment (cyberwolf) [inline tests/features/change_tracking_workflow.feature]: I should see **a** "div.ICE-Tracking" element

File: tests/features/change_tracking_workflow.feature [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+      | title            | Lorem ipsum dolor sit amet                                      |
+      | body             | <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p> |
+      | moderation state | draft                                                           |
+    When I click "Edit draft"
+    And I fill in "Body" with "<html>"
+    And I press "Save"
+    Then the response should contain "<expected>"
+    And I should not see the following warning messages:
+      | warning messages | <strong>The change tracking is activated on some fields of this "Basic page" content</strong>.<br /> <small>Please accept or reject tracked changes before setting the content state to validated or published.</small> |
+    And I should not see an "div.ICE-Tracking" element
```
Comment (cyberwolf) [inline tests/features/change_tracking_workflow.feature]: I should not see a "div.ICE-Tracking" element

File: tests/features/change_tracking_workflow.feature [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+    And the following languages are available:
+      | languages |
+      | en        |
+      | fr        |
+    When I go to "admin/config/content/multisite_wysiwyg/workbench"
+    And I check the box "edit-nexteuropa-editorial-tracking-wbm-states-validated"
+    And I check the box "edit-nexteuropa-editorial-tracking-wbm-states-published"
+    And I press "Save configuration"
+    Then I should see the success message "The configuration options have been saved."
+    And I am viewing a multilingual "page" content:
```
Comment (cyberwolf) [inline tests/features/change_tracking_workflow.feature]: Use When here because the previous line is a Then, but this line does not verify an observable outcome

File: tests/features/change_tracking_workflow.feature (hunk 24 above)
Comment (gillesdeudon) [inline tests/features/change_tracking_workflow.feature]: Here, it is correct. It is the draft version of the content.

File: tests/features/change_tracking_workflow.feature (hunk 17 above)
Comment (gillesdeudon) [inline tests/features/change_tracking_workflow.feature]: I take into account the remark... but for the next time. On this ticket, I already spent a lot of time and I fear that kind of improvements consume some time in trial and error steps before having the perfect display.

**Comment by cyberwolf** [comment]
👍

**Comment by voidtek** [comment]
:+1:

**Comment by cyberwolf** [comment]
👍

**Comment by voidtek** [comment]
:+1:

**Comment by netlooker** [comment]
👍

**Comment by cyberwolf** [comment]
:+1

**Comment by cyberwolf** [comment]
👍
